### ISSUE TITLE
{strip} can produce invalid HTML tags

### ISSUE THREAD

**Issue by rafal-zajdel** [smarty-php/smarty#764]
I've found out that `{strip}` removes all whitespace characters between attribute and value pairs while used on multiline HTML tag with Smarty variable inside.

My Smarty version is 3.1.45. Template code

```Smarty
{assign var='extension' value='png'}

{strip}
    <img src="image.{$extension}"
         alt="image"
    />
{/strip}

{strip}
    <img src="image.png"
         alt="image"
    />
{/strip}
```

produces

```html
<img src="image.png"alt="image"/>

<img src="image.png" alt="image" />
```

First image tag is invalid but browsers seem to handle this well.

**Comment by wisskid** [comment]
It may not be what you expected, but this behavior is literally in the [docs](https://smarty-php.github.io/smarty/designers/language-builtin-functions/language-function-strip.html). Even tho this function was written with HTML content in mind, Smarty does not understand HTML, it just understands the parts between `{` and `}`. Changing this would cause a lot of backwards compatibility problems, so I'm not inclined to do so.

**Comment by rafal-zajdel** [comment]
Thank you for the clarification. I saw special treatment of some tags like `<script>` or `<pre>` and I thought Smarty is more HTML-aware. Could you suggest any solution that will allow multiline HTML tags with Smarty code inside of `{strip}`? I can't think of any other solution than simple replacement of newlines by spaces within HTML tags.

**Comment by wisskid** [comment]
Yes, that’s what I would do. You can write a pugin block that does exactly that. First replace newlines by spaces, then truncate repeated spaces to a single space.

I don’t know your use-case, but if all you want is to reduce the size of the html, something like Apache’s mod_deflate might be a more powerful option.

**Comment by rafal-zajdel** [comment]
I use `{strip}` to simplify work with HTML tags that have `display: inline-block;` style and also to optimize text-to-code ratio. I've prepared very simple block plugin just to test some concepts:

```php
function smarty_block_force_single_line_tags($params, $content, $template, &$repeat)
{
    if ($repeat === true) {
        return false;
    }

    return preg_replace('`[\r\n]+`u', ' ', $content);
}
```

and noticed that Smarty code

```Smarty
{assign var='extension' value='png'}

{strip}
	{force_single_line_tags}
		<img src="image.{$extension}"
			 alt="image"
		/>
	{/force_single_line_tags}
{/strip}
```

still produces invalid HTML output:

```html
<img src="image.png"alt="image"/>
```

It looks to me that the only solution to my problem (other than rewriting the code) would be creating my own block plugin that would process HTML to remove unnecessary whitespaces within HTML tags and around them. This kind of processing seems to be too complicated to run on every request as block plugin output seems to be uncacheable. To make this block plugin trustworthy I would have to ensure that attribute values and also contents of tags such as `<style>`, `<script>`, `<pre>`, `<code>`, `<textarea>` etc. are kept unchanged which would require some nontrivial and costly PHP code execution, which would have negative impact on TTFB :)
